### ISSUE TITLE
Cannot access supertrait items in trait definition

### ISSUE THREAD

**Issue by djc** [rust-lang/rust#72192]
I tried this code:

```rust
trait Trait: SuperTrait {
    fn array() -> [u8; <Self as SuperTrait>::SIZE];
}

trait SuperTrait {
    const SIZE: usize;
}
```

https://play.rust-lang.org/?version=nightly&mode=debug&edition=2018&gist=9d962e9ed3712729d2d9150df48ecddf

I expected to see this happen: compiled without errors.

Instead, this happened:

```rust
error[E0277]: the trait bound `Self: SuperTrait` is not satisfied
 --> src/lib.rs:2:24
  |
2 |     fn array() -> [u8; <Self as SuperTrait>::SIZE];
  |                        ^^^^^^^^^^^^^^^^^^^^^^^^^^ - help: consider further restricting `Self`: `where Self: SuperTrait`
  |                        |
  |                        the trait `SuperTrait` is not implemented for `Self`
...
6 |     const SIZE: usize;
  |     ------------------ required by `SuperTrait::SIZE`
```

I'm not sure whether rustc is incapable of accessing the supertrait items (because Chalk, I guess), or whether it's erroneously detecting an error.

**Comment by jonas-schievink** [comment]
Since `fn array() -> usize { <Self as SuperTrait>::SIZE }` works fine, this is a duplicate of https://github.com/rust-lang/rust/issues/43408
